src/app/clusters/general-commissioning-server/general-commissioning-server.cpp

```diff
@@ -328,23 +521,105 @@ bool emberAfGeneralCommissioningClusterSetRegulatoryConfigCallback(app::CommandH
     return true;
 }
 
+bool emberAfGeneralCommissioningClusterSetTCAcknowledgementsCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::GeneralCommissioning::Commands::SetTCAcknowledgements::DecodableType & commandData)
+{
+#if CHIP_CONFIG_TC_REQUIRED
+    MATTER_TRACE_SCOPE("SetTCAcknowledgements", "GeneralCommissioning");
+
+    auto & failSafeContext                                        = Server::GetInstance().GetFailSafeContext();
+    TermsAndConditionsProvider * const termsAndConditionsProvider = Server::GetInstance().GetTermsAndConditionsProvider();
+
+    Optional<TermsAndConditions> acceptedTermsAndConditions = Optional<TermsAndConditions>({
+        .value   = commandData.TCUserResponse,
+        .version = commandData.TCVersion,
+    });
+
+    Commands::SetTCAcknowledgementsResponse::Type response;
+    response.errorCode = CheckTermsAndConditionsAcknowledgementsState(termsAndConditionsProvider, acceptedTermsAndConditions);
+
+    if (CommissioningErrorEnum::kOk == response.errorCode)
+    {
+        CheckSuccess(termsAndConditionsProvider->SetAcceptance(acceptedTermsAndConditions), Failure);
+
+        if (failSafeContext.IsFailSafeArmed())
+        {
+            failSafeContext.SetUpdateTermsAndConditionsHasBeenInvoked();
+        }
+        else
+        {
+            CheckSuccess(termsAndConditionsProvider->CommitAcceptance(), Failure);
```

Comment:
Why is this allowed to be called (and immediately committed) outside failsafe?

Reply:
Outside of the failsafe context, I believe we should be able to accept updated terms and conditions. I thought the intention of failsafe context was to establish a failsafe context where you could change things and they are not committed until the failsafe is closed or commissioning complete. I don't believe we made any statements that this command may only be received while in the failsafe context.

Reply:
Every command of General Commissioning cluster can only be handled during fail-same. But overall, it's odd whenever there is any state that can be reverted by fail-safe under some situations, but not others.

Reply:
That's not correct. The SetRegulatoryConfig can be called outside of a fail-safe context. It also isn't rolled back. Should we open an issue to help clarify the fail-safe context handling?

Reply:
Yes, please. 

Regulatory config should also be part of the fail-safe.

Reply:
Issue filed: https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/10686